lucene/core/src/java/org/apache/lucene/index/DocumentsWriterFlushControl.java

```diff
@@ -380,52 +368,35 @@ DocumentsWriterPerThread nextPendingFlush() {
       fullFlush = this.fullFlush;
       numPending = this.numPending;
     }
-    if (numPending > 0 && !fullFlush) { // don't check if we are doing a full flush
-      final int limit = perThreadPool.getActiveThreadStateCount();
-      for (int i = 0; i < limit && numPending > 0; i++) {
-        final ThreadState next = perThreadPool.getThreadState(i);
-        if (next.flushPending) {
-          final DocumentsWriterPerThread dwpt = tryCheckoutForFlush(next);
-          if (dwpt != null) {
-            return dwpt;
+    if (numPending > 0 && fullFlush == false) { // don't check if we are doing a full flush
+      for (final DocumentsWriterPerThread next : perThreadPool) {
+        if (next.isFlushPending()) {
+          if (next.tryLock()) {
+            try {
+              if (perThreadPool.isRegistered(next)) {
+                return checkOutForFlush(next);
+              }
+            } finally {
+              next.unlock();
```

Comment:
Is it intentional that we are `unlock`ing `next` even after returning it from `checkOutForFlush(next)`?  Shouldn't it remain locked until flush finishes?

Reply:
so far it's intentional that we are checking out the DWPT of the pool and unlock it. The reason is that others might block on it to check it out as well but then won't succeed since it's not part of the pool anymore. I can take another iteration here and make sure it's locked until flushed. I wonder if we should do this after the fact.

Reply:
I was just worried that unlocking on return was not intended -- sometimes people think `finally` clauses are not executed on `return` ;)

If it is intentional that's great.

Reply:
I think this should release again. The way of use of finally here is my favourite use of finally. You can make very elegant methods body's with returns and still have finally executed.